File: BOJ/Algorithms/Graph/DFS_BFS/11725.cpp
#include <iostream>
#include <vector>
#define MAX 100001
using namespace std;

bool Visit[MAX];
int parent[MAX];
vector<int> Tree[MAX];

void DFS(int V){
    Visit[V] = true;
    for(int i = 0; i < Tree[V].size(); i++){
        int next = Tree[V][i];
        if(!Visit[next]){
            parent[next] = V;
            DFS(next);
        }
    }
}

int main(){
    int N;
    cin >> N;
    for(int i = 0; i < N - 1; i++){
        int A, B;
        cin >> A >> B;
        Tree[A].push_back(B);
        Tree[B].push_back(A);
    }
    DFS(1);
    for(int i = 2 ; i <= N; i++) cout << parent[i] << '\n';
}

File: BOJ/Algorithms/Graph/DFS_BFS/11725.py
import sys ; input = lambda : sys.stdin.readline().rstrip()
sys.setrecursionlimit(10**6)
def DFS(V):
    visit[V] = True
    for i in range(len(Tree[V])):
        next = Tree[V][i]
        if visit[next] == 0:
            parent[next] = V
            DFS(next)

N = int(input())

Tree = [[] for _ in range(N+1)]
visit = [0 for _ in range(N+1)]
parent = [0 for _ in range(N+1)]

for i in range(N-1):
    A, B = map(int, input().split())
    Tree[A].append(B)
    Tree[B].append(A)

DFS(1)
for i in range(2, N+1):
    print(parent[i])